src/meta/src/hummock/compaction_schedule_policy.rs

```diff
@@ -354,12 +436,99 @@ impl CompactionSchedulePolicy for ScoredPolicy {
         self.score_to_compactor.len()
     }
 
+    fn suggest_scale_policy(&self) -> ScalePolicy {
+        match &self.state {
+            CompactorState::Busy(_) => return ScalePolicy::ScaleOut,
+            CompactorState::Burst(_) => (),
+            CompactorState::Idle(last_idle_time) => {
+                if last_idle_time.elapsed().as_secs() > MAX_IDLE_TIME
+                    && self.score_to_compactor.len() > 1
+                {
+                    let decrease_core = self
+                        .score_to_compactor
+                        .values()
+                        .map(|compactor| compactor.total_cpu_core as u64)
+                        .min()
+                        .unwrap_or(0);
+                    return ScalePolicy::ScaleIn(decrease_core);
+                }
+            }
+        }
+        ScalePolicy::NoChange
+    }
+
     fn max_concurrent_task_num(&self) -> usize {
         self.score_to_compactor
             .values()
             .map(|c| c.max_concurrent_task_number() as usize)
             .sum()
     }
+
+    /// `refresh_state` check the state of all `Compactors` and calculate a new state based on the
+    /// state and duration
+    /// The state of `CompactorPolicy` must maintain `MIN_LAST_STATE_TIME` before it will change, to
+    /// ensure that the state changes do not jitter frequently.
+    fn refresh_state(&mut self) {
+        let idle_count = self
+            .score_to_compactor
+            .values()
+            .filter(|compactor| !compactor.is_busy())
+            .count();
+
+        if idle_count == 0 {
+            match self.state {
+                CompactorState::Idle(last_update) => {
+                    if last_update.elapsed().as_secs() > MIN_LAST_STATE_TIME {
+                        self.state = CompactorState::Burst(Instant::now());
+                    }
+                }
+                CompactorState::Burst(last_update) => {
+                    if last_update.elapsed().as_secs() > MIN_LAST_STATE_TIME {
+                        self.state = CompactorState::Busy(Instant::now());
```

Comment:
The logic in every match arm is very similar. Can we abstract a `upgrade_state` and `downgrade_state` function on `CompactorState`?